File: onebot_manager.py
import asyncio
import inspect
import json
import logging
import re
import traceback
from typing import Dict, Set, Optional, Any, Callable, Union, AsyncGenerator
from uuid import uuid4
from fastapi import WebSocket, WebSocketDisconnect

logger = logging.getLogger("onebot")


def extract_token(auth_header: Optional[str]) -> Optional[str]:
    """从 Authorization 头部提取 token"""
    if not auth_header:
        return None

    match = re.compile(r'^(?:token|bearer)\s+(\S+)$', re.IGNORECASE).match(auth_header.strip())
    return match.group(1) if match else None


# noinspection PyBroadException
class OneBotConnectionManager:
    def __init__(self, token: Optional[str] = None):
        self.token = token
        self.active_connections: Dict[int, WebSocket] = {}  # self_id (int) -> WebSocket
        self.connection_states: Dict[WebSocket, Dict[str, Any]] = {}  # 连接状态存储
        self.pending_actions: Dict[str, Union[asyncio.Future, asyncio.Queue]] = {}
        self.message_handlers: Set[Callable[[Dict[str, Any]], Any]] = set()

    async def connect(self, websocket: WebSocket):
        authorization = websocket.headers.get('Authorization', None)

        if not await self.authenticate(websocket, authorization):
            return

        await websocket.accept()

        try:
            while True:
                message = await websocket.receive_json()
                await self.handle_message(websocket, message)
        except WebSocketDisconnect:
            await self.disconnect(websocket)
        except Exception as e:
            print(f"WebSocket error: {e}")
            await self.disconnect(websocket)

    async def authenticate(self, websocket: WebSocket, auth_header: Optional[str] = None) -> bool:
        """验证连接Token"""
        if not self.token:
            return True

        token = extract_token(auth_header)
        if token != self.token:
            # print('Invalid token:', token)
            await websocket.close(code=1008, reason="Invalid token")
            return False
        return True

    async def register_connection(self, websocket: WebSocket, self_id: int) -> bool:
        """注册已验证的连接"""
        if self_id in self.active_connections:
            await websocket.close(code=1008, reason=f"Duplicate self_id: {self_id}")
            return False

        self.active_connections[self_id] = websocket
        self.connection_states[websocket] = {"self_id": self_id, "authenticated": True}
        logger.info(f"OneBot connected: {self_id}")
        return True

    async def disconnect(self, websocket: WebSocket) -> None:
        """断开连接并清理资源"""
        state = self.connection_states.get(websocket)
        if state and "self_id" in state:
            self_id = state["self_id"]
            self.active_connections.pop(self_id, None)
            logger.info(f"OneBot disconnected: {self_id}")
        self.connection_states.pop(websocket, None)

        try:
            await websocket.close()
        except:
            pass

    def get_first_self_id(self):
        return next(iter(self.active_connections.keys()))

    async def call_action(self, action, params, self_id=None, timeout: float = 60.0):
        if not self.active_connections:
            raise Exception("No active OneBot connections")

        if self_id is None:
            if len(self.active_connections) == 1:
                self_id = next(iter(self.active_connections.keys()))
            else:
                raise Exception("self_id required when multiple bots connected")

        result = await self.send_action(
            self_id=self_id,
            action=action,
            params=params,
            timeout=timeout
        )

        if result.get('status') == 'ok':
            return result.get('data')

        return result

    def cancel_action(self, echo: str):
        """取消一个正在等待响应的 action"""
        future = self.pending_actions.pop(echo, None)
        if future and not future.done():
            future.cancel()
            print(f"OneBot action cancelled (echo: {echo})")

    # noinspection PyAsyncCall
    async def send_action(self, self_id: int, action: str, params: Dict[str, Any], timeout: float = 60.0) -> Any:
        """发送API动作并等待响应"""
        if self_id not in self.active_connections:
            raise ConnectionError(f"No active connection for self_id: {self_id}")

        echo = str(uuid4())
        future = asyncio.get_event_loop().create_future()
        self.pending_actions[echo] = future

        try:
            await self.active_connections[self_id].send_json({
                "action": action,
                "params": params,
                "echo": echo
            })

            result_data = await asyncio.wait_for(future, timeout)

            return result_data
        except asyncio.TimeoutError:
            self.pending_actions.pop(echo, None)
            raise TimeoutError("Action timed out")
        except asyncio.CancelledError:
            self.pending_actions.pop(echo, None)
            raise
        except ActionFailed as e:
            self.pending_actions.pop(echo, None)
            raise e
        except Exception as e:
            traceback.print_exc()
            self.pending_actions.pop(echo, None)
            raise e

    async def call_stream_action(self, action: str, params: Dict[str, Any],
                                 self_id: Optional[int] = None,
                                 timeout: float = 30.0) -> AsyncGenerator[Dict[str, Any], None]:
        """发送流式API动作并异步生成响应"""
        if not self.active_connections:
            raise Exception("No active OneBot connections")

        if self_id is None:
            if len(self.active_connections) == 1:
                self_id = next(iter(self.active_connections.keys()))
            else:
                raise Exception("self_id required when multiple bots connected")

        echo = str(uuid4())
        queue = asyncio.Queue()
        self.pending_actions[echo] = queue

        try:
            await self.active_connections[self_id].send_json({
                "action": action,
                "params": params,
                "echo": echo
            })

            while True:
                try:
                    data = await asyncio.wait_for(queue.get(), timeout)
                    if data.get("data", {}).get("type") == "response":
                        break  # 遇到最终响应时结束
                    yield data
                    if data.get("data", {}).get("type") == "error":
                        raise Exception(f"Stream action failed: {data.get('data', {})}")
                except asyncio.TimeoutError:
                    raise TimeoutError("Stream action timed out")
                except Exception as e:
                    raise e

        finally:
            # noinspection PyAsyncCall
            self.pending_actions.pop(echo, None)

    async def handle_message(self, websocket: WebSocket, message: dict) -> None:
        """处理接收到的消息"""
        try:
            data = message

            # 处理连接生命周期事件
            if data.get("post_type") == "meta_event" and data.get("meta_event_type") == "lifecycle":
                state = self.connection_states.get(websocket, {})
                if not state.get("authenticated"):
                    # 从WebSocket headers中获取Authorization
                    auth_header = websocket.headers.get("Authorization")
                    if not await self.authenticate(websocket, auth_header):
                        return
                    if not await self.register_connection(websocket, int(data["self_id"])):
                        return

            # 处理API响应
            if "echo" in data and isinstance(data["echo"], str):
                echo = data["echo"]
                if echo in self.pending_actions:
                    # 处理普通动作响应
                    if echo in self.pending_actions and isinstance(self.pending_actions[echo], asyncio.Future):
                        future = self.pending_actions[echo]
                        if "status" in data and data["status"] == "failed":
                            future.set_exception(ActionFailed(data))
                        else:
                            future.set_result(data)
                        return

                    # 处理流式动作响应
                    if echo in self.pending_actions and isinstance(self.pending_actions[echo], asyncio.Queue):
                        queue = self.pending_actions[echo]
                        await queue.put(data)
                        return

            # 处理事件推送
            if "post_type" in data:
                for handler in list(self.message_handlers):
                    try:
                        if inspect.iscoroutinefunction(handler):
                            await handler(data)
                        else:
                            handler(data)
                    except Exception as e:
                        logger.error(f"Error in message handler: {e}", exc_info=True)

        except json.JSONDecodeError:
            logger.error(f"Invalid JSON message: {message}")

    def add_message_handler(self, handler: Callable[[Dict[str, Any]], Any]) -> Callable[[], None]:
        """添加消息处理器"""
        self.message_handlers.add(handler)
        return lambda: self.message_handlers.discard(handler)

    async def broadcast_event(self, event: Dict[str, Any]) -> None:
        """向所有连接的客户端广播事件"""
        for ws in list(self.active_connections.values()):
            try:
                await ws.send_json(event)
            except:
                await self.disconnect(ws)


class ActionFailed(Exception):
    def __init__(self, data: Dict[str, Any]):
        self.data = data
        super().__init__(f"Action failed: {data.get('message', data.get('wording', 'Unknown error'))}")

File: server/src/onebot_manager.ts
import { randomUUID } from 'node:crypto';
import type { WebSocket } from '@fastify/websocket';
import type { FastifyRequest } from 'fastify';
import { extractToken } from './utils.js';

/**
 * Action 失败异常
 */
export class ActionFailed extends Error {
  data: Record<string, unknown>;

  constructor(data: Record<string, unknown>) {
    super(`Action failed: ${String(data['message'] ?? data['wording'] ?? 'Unknown error')}`);
    this.name = 'ActionFailed';
    this.data = data;
  }
}

/**
 * 待处理 action 条目
 */
interface PendingActionEntry {
  resolve: (value: unknown) => void;
  reject: (reason: unknown) => void;
}

/**
 * 流式 action 队列条目
 */
interface StreamQueueEntry {
  data: Record<string, unknown>;
  resolve: () => void;
}

/**
 * 流式 action 状态
 */
class StreamActionState {
  queue: StreamQueueEntry[] = [];
  waiters: { resolve: (entry: StreamQueueEntry) => void; reject: (err: Error) => void }[] = [];
  done = false;

  push(data: Record<string, unknown>): void {
    const entry: StreamQueueEntry = {
      data,
      resolve: () => {},
    };
    if (this.waiters.length > 0) {
      const waiter = this.waiters.shift()!;
      waiter.resolve(entry);
    } else {
      this.queue.push(entry);
    }
  }

  async pop(timeout: number): Promise<StreamQueueEntry> {
    if (this.queue.length > 0) {
      return this.queue.shift()!;
    }
    if (this.done) {
      throw new Error('Stream ended');
    }
    return new Promise<StreamQueueEntry>((resolve, reject) => {
      const timer = setTimeout(() => {
        const idx = this.waiters.indexOf({ resolve, reject } as any);
        if (idx >= 0) this.waiters.splice(idx, 1);
        reject(new Error('Stream action timed out'));
      }, timeout * 1000);
      this.waiters.push({
        resolve: (entry: StreamQueueEntry) => {
          clearTimeout(timer);
          resolve(entry);
        },
        reject: (err: Error) => {
          clearTimeout(timer);
          reject(err);
        },
      });
    });
  }

  finish(): void {
    this.done = true;
    for (const waiter of this.waiters) {
      waiter.reject(new Error('Stream ended'));
    }
    this.waiters = [];
  }
}

/**
 * OneBot WebSocket 连接管理器
 */
export class OneBotConnectionManager {
  token: string | null;
  activeConnections: Map<number, WebSocket> = new Map(); // self_id (int) -> WebSocket
  connectionStates: Map<WebSocket, Record<string, unknown>> = new Map();
  pendingActions: Map<string, PendingActionEntry> = new Map();
  messageHandlers: Set<(data: Record<string, unknown>) => Promise<void> | void> = new Set();
  _streamStates: Map<string, StreamActionState> = new Map();
  _streamEchoes: Set<string> = new Set();

  constructor(token: string | null = null) {
    this.token = token;
  }

  async connect(websocket: WebSocket, request: FastifyRequest): Promise<void> {
    const authorization = request.headers['authorization'] ?? null;

    if (!await this.authenticate(websocket, authorization as string | null | undefined)) {
      return;
    }

    // @fastify/websocket 已自动 accept 连接
    // 使用事件监听方式处理消息（ws 库事件驱动模式）
    const onMessage = (rawData: Buffer | string) => {
      try {
        const messageStr = typeof rawData === 'string' ? rawData : rawData.toString();
        const data = JSON.parse(messageStr) as Record<string, unknown>;
        // 异步处理但不阻塞事件循环
        this.handleMessage(websocket, data).catch((err) => {
          console.error(`WebSocket message handler error: ${err}`);
        });
      } catch (err) {
        console.error(`WebSocket message parse error: ${err}`);
      }
    };

    const onClose = () => {
      this.disconnect(websocket).catch((err) => {
        console.error(`WebSocket disconnect error: ${err}`);
      });
    };

    const onError = (err: Error) => {
      console.error(`WebSocket error: ${err}`);
    };

    websocket.on('message', onMessage);
    websocket.on('close', onClose);
    websocket.on('error', onError);
  }

  async authenticate(websocket: WebSocket, authHeader: string | null | undefined): Promise<boolean> {
    /* 验证连接Token */
    if (!this.token) return true;

    const token = extractToken(authHeader);
    if (token !== this.token) {
      console.log('Invalid token:', token);
      websocket.close(1008, 'Invalid token');
      return false;
    }
    return true;
  }

  async registerConnection(websocket: WebSocket, selfId: number): Promise<boolean> {
    /* 注册已验证的连接 */
    if (this.activeConnections.has(selfId)) {
      websocket.close(1008, `Duplicate self_id: ${selfId}`);
      return false;
    }

    this.activeConnections.set(selfId, websocket);
    this.connectionStates.set(websocket, { self_id: selfId, authenticated: true });
    console.log(`OneBot connected: ${selfId}`);
    return true;
  }

  async disconnect(websocket: WebSocket): Promise<void> {
    /* 断开连接并清理资源 */
    const state = this.connectionStates.get(websocket);
    if (state && 'self_id' in state) {
      const selfId = state['self_id'] as number;
      this.activeConnections.delete(selfId);
      console.log(`OneBot disconnected: ${selfId}`);
    }
    this.connectionStates.delete(websocket);

    try {
      websocket.close();
    } catch {
      // 忽略关闭错误
    }
  }

  getFirstSelfId(): number | null {
    const first = this.activeConnections.keys().next();
    return first.done ? null : first.value;
  }

  async callAction(action: string, params: Record<string, unknown>, selfId: number | null = null, timeout: number = 60): Promise<unknown> {
    if (this.activeConnections.size === 0) {
      throw new Error('No active OneBot connections');
    }

    if (selfId === null) {
      if (this.activeConnections.size === 1) {
        selfId = this.getFirstSelfId()!;
      } else {
        throw new Error('self_id required when multiple bots connected');
      }
    }

    const result = await this.sendAction(selfId, action, params, timeout);

    if (typeof result === 'object' && result !== null && 'status' in (result as Record<string, unknown>)) {
      const r = result as Record<string, unknown>;
      if (r['status'] === 'ok') {
        return r['data'];
      }
    }

    return result;
  }

  cancelAction(echo: string): void {
    /* 取消一个正在等待响应的 action */
    const pending = this.pendingActions.get(echo);
    if (pending) {
      this.pendingActions.delete(echo);
      console.log(`OneBot action cancelled (echo: ${echo})`);
    }
  }

  async sendAction(selfId: number, action: string, params: Record<string, unknown>, timeout: number = 60): Promise<unknown> {
    /* 发送API动作并等待响应 */
    const ws = this.activeConnections.get(selfId);
    if (!ws) {
      throw new Error(`No active connection for self_id: ${selfId}`);
    }

    const echo = randomUUID();
    const { promise, resolve, reject } = Promise.withResolvers<unknown>();
    this.pendingActions.set(echo, { resolve, reject });

    try {
      ws.send(JSON.stringify({
        action,
        params,
        echo,
      }));

      const result = await promise as Record<string, unknown>;
      return result;
    } catch (err) {
      if (err instanceof Error) {
        throw err;
      }
      throw new Error(String(err));
    } finally {
      this.pendingActions.delete(echo);
    }
  }

  async callStreamAction(
    action: string,
    params: Record<string, unknown>,
    selfId: number | null = null,
    timeout: number = 30,
  ): Promise<AsyncGenerator<Record<string, unknown>, void, unknown>> {
    /* 发送流式API动作并异步生成响应 */
    if (this.activeConnections.size === 0) {
      throw new Error('No active OneBot connections');
    }

    if (selfId === null) {
      if (this.activeConnections.size === 1) {
        selfId = this.getFirstSelfId()!;
      } else {
        throw new Error('self_id required when multiple bots connected');
      }
    }

    const ws = this.activeConnections.get(selfId);
    if (!ws) {
      throw new Error(`No active connection for self_id: ${selfId}`);
    }

    const echo = randomUUID();
    const streamState = new StreamActionState();
    this.pendingActions.set(echo, {
      resolve: () => {},
      reject: () => {},
    });

    // 存储流状态到单独的 map 中
    this._streamStates.set(echo, streamState);

    // 保存 stream echo 引用
    this._streamEchoes.add(echo);

    try {
      ws.send(JSON.stringify({
        action,
        params,
        echo,
      }));

      const generator = async function* (
        mgr: OneBotConnectionManager,
        streamEcho: string,
        state: StreamActionState,
        timeoutSec: number,
      ) {
        try {
          while (true) {
            const entry = await state.pop(timeoutSec);
            const data = entry.data;

            // 检查是否到达最终响应
            if (data['data'] && typeof data['data'] === 'object') {
              const innerData = data['data'] as Record<string, unknown>;
              if (innerData['type'] === 'response') {
                break; // 遇到最终响应时结束
              }
              if (innerData['type'] === 'error') {
                throw new Error(`Stream action failed: ${JSON.stringify(innerData)}`);
              }
            }

            yield data;
          }
        } finally {
          mgr._streamStates.delete(streamEcho);
          mgr._streamEchoes.delete(streamEcho);
          mgr.pendingActions.delete(streamEcho);
        }
      };

      return generator(this, echo, streamState, timeout);
    } catch (err) {
      this._streamStates.delete(echo);
      this._streamEchoes.delete(echo);
      this.pendingActions.delete(echo);
      throw err;
    }
  }

  async handleMessage(websocket: WebSocket, message: Record<string, unknown>): Promise<void> {
    /* 处理接收到的消息 */
    try {
      const data = message;

      // 处理连接生命周期事件
      if (data['post_type'] === 'meta_event' && data['meta_event_type'] === 'lifecycle') {
        const state = this.connectionStates.get(websocket) ?? {};
        if (!state['authenticated']) {
          if (!await this.authenticate(websocket, null)) {
            return;
          }
          if (!await this.registerConnection(websocket, Number(data['self_id']))) {
            return;
          }
        }
      }

      // 处理API响应
      if (typeof data['echo'] === 'string') {
        const echo = data['echo'] as string;

        // 先检查是否是流式 action 响应
        if (this._streamStates.has(echo)) {
          const state = this._streamStates.get(echo)!;
          if ('status' in data && data['status'] === 'failed') {
            state.push({ data: { type: 'error', ...data } as unknown as Record<string, unknown>, resolve: () => {} });
            state.finish();
          } else {
            state.push(data);
          }
          return;
        }

        const pending = this.pendingActions.get(echo);
        if (pending) {
          if ('status' in data && data['status'] === 'failed') {
            pending.reject(new ActionFailed(data));
          } else {
            pending.resolve(data);
          }
          return;
        }
      }

      // 处理事件推送
      if ('post_type' in data) {
        for (const handler of this.messageHandlers) {
          try {
            const result = handler(data);
            if (result instanceof Promise) {
              await result;
            }
          } catch (err) {
            console.error(`Error in message handler: ${err}`);
          }
        }
      }
    } catch (err) {
      console.error(`Invalid message: ${err}`);
    }
  }

  addMessageHandler(handler: (data: Record<string, unknown>) => Promise<void> | void): () => void {
    /* 添加消息处理器 */
    this.messageHandlers.add(handler);
    return () => {
      this.messageHandlers.delete(handler);
    };
  }

  async broadcastEvent(event: Record<string, unknown>): Promise<void> {
    /* 向所有连接的客户端广播事件 */
    for (const ws of this.activeConnections.values()) {
      try {
        ws.send(JSON.stringify(event));
      } catch {
        await this.disconnect(ws);
      }
    }
  }
}